[shaderc] Shaderc doesn't install a "config" CMake file, so it can't be found with find_package after a successful install

****

For example, if you install the package glslang, you get vcpkg\installed\x64-windows\share\glslang\glslang-config.cmake (and a bunch of other related cmake magic files in that directory.)  But if you install shaderc, it does build the .lib file, and the install process completes successfully, but no similar shaderc-config is generated/exported.  As far as I can tell, you need to manually find the library and include files as a result.

Upstream shaderc is a bit eccentric.  It still uses CMakee 2.x, and the build system hasn't been updated for "modern" package manager friendly install stuff:  https://github.com/google/shaderc/blob/main/CMakeLists.txt#L15

Tring to add install(TARGETS shaderc shaderc_util EXPORT shaderc-config ...) to the CMakeLists.txt resulted in a configure time error

CMake Error in libshaderc/CMakeLists.txt: Target "shaderc" INTERFACE_INCLUDE_DIRECTORIES property contains path:which is prefixed in the source directory.

After some googling, I have been fiddling with $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include/> vs  $<INSTALL_INTERFACE:include/> in the target_include_directories, which make the configure time CMake error go away, but so far I can only get that to build the package if I also patch the source being built to change the include statements because I apparently haven't made a generator expression that exactly matches what the source is expecting.

**Proposed solution**
???  Hope somebody who knows more CMake than me can fix it.

I have been working on hacking up a CMake build system that works, but I have been cargo-culting some of the install stuff that I don't really understand very well.  And so far I have only tried to install it on a Windows machine, without any idea about testing some of the other configurations that shaderc does weird build stuff with like supporting Android with find_package being different from find_host_package, or all of the stuff they have behind if(NOT MSVC) on their windows build.

It's possible that patching the build in the vcpkg port is just the wrong approach, and the upstream build needs some love and modernization, but I'm not sure exactly what should be requested on that route since I don't really understand the CMake exports + installs stuff, and what the implications might be for other package managers besides vcpkg wanting to consume shaderc.

Oh, also, shaderc has stuff like glslang in a third party directory.  Obviously, the ideal result should consume vcpkg's libraries, rather than rebuilding anything the vcpkg already provides.

**Describe alternatives you've considered**

Seriously considered moving to a remote cabin without electricity.

**Additional context**

If I didn't have electricity, I wouldn't be able to do anything related to CMake.  That sounds nice.

**Additional Additional context**

If you are using MSVC and try to set the shaderc_util library to be SHARED like this:

add_library(shaderc SHARED ${SHADERC_SOURCES})

you get a bunch of errors about runtime library mismatches

MachineIndependentd.lib(IntermTraverse.cpp.obj):-1: error: LNK2038: mismatch detected for 'RuntimeLibrary': value 'MDd_DynamicDebug' doesn't match value 'MTd_StaticDebug' in shaderc.cc.obj

So, there be dragons.  I haven't fully unwound what all is happening in shaderc's cmake modules to see if they are doing something wildly different from normal.  Just throwing it out there.

--- a/ports/shaderc/fix-export-cmakefiles.patch
+++ b/ports/shaderc/fix-export-cmakefiles.patch
@@ -1,0 +1,93 @@
+diff --git a/libshaderc/CMakeLists.txt b/libshaderc/CMakeLists.txt
+index f3277f0..5fd58ef 100644
+--- a/libshaderc/CMakeLists.txt
++++ b/libshaderc/CMakeLists.txt
+@@ -28,7 +28,9 @@ if (NOT BUILD_SHARED_LIBS)
+ add_library(shaderc STATIC ${SHADERC_SOURCES})
+ shaderc_default_compile_options(shaderc)
+ target_include_directories(shaderc
+-    PUBLIC include
++    PUBLIC
++    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include/>
++    $<INSTALL_INTERFACE:include/>
+     PRIVATE ${glslang_SOURCE_DIR}
+             ${SPIRV-Headers_SOURCE_DIR}/include)
+ set(shaderc_install_target shaderc)
+@@ -37,7 +39,9 @@ else()
+ add_library(shaderc_shared SHARED ${SHADERC_SOURCES})
+ shaderc_default_compile_options(shaderc_shared)
+ target_include_directories(shaderc_shared
+-    PUBLIC include
++    PUBLIC
++    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include/>
++    $<INSTALL_INTERFACE:include/>
+     PRIVATE ${glslang_SOURCE_DIR}
+             ${SPIRV-Headers_SOURCE_DIR}/include)
+ target_compile_definitions(shaderc_shared
+@@ -59,10 +63,14 @@ if(SHADERC_ENABLE_INSTALL)
+     DESTINATION
+       ${CMAKE_INSTALL_INCLUDEDIR}/shaderc)
+ 
+-  install(TARGETS ${shaderc_install_target}
++  install(TARGETS ${shaderc_install_target} EXPORT unofficial-shadercTargets
+     LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+     RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+     ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
++  install(EXPORT unofficial-shadercTargets
++    NAMESPACE unofficial::shaderc::
++    DESTINATION share/unofficial-shaderc
++  )
+ endif(SHADERC_ENABLE_INSTALL)
+ 
+ find_package(Threads)
+@@ -79,6 +87,11 @@ else()
+ target_link_libraries(shaderc_shared PRIVATE ${SHADERC_LIBS})
+ endif()
+ 
++file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/unofficial-shaderc-config.cmake.in"
++[[include("${CMAKE_CURRENT_LIST_DIR}/unofficial-shadercTargets.cmake")]])
++configure_file("${CMAKE_CURRENT_BINARY_DIR}/unofficial-shaderc-config.cmake.in" "${CMAKE_CURRENT_BINARY_DIR}/unofficial-shaderc-config.cmake" @ONLY)
++install(FILES ${CMAKE_CURRENT_BINARY_DIR}/unofficial-shaderc-config.cmake DESTINATION share/unofficial-shaderc)
++
+ shaderc_add_tests(
+   TEST_PREFIX shaderc
+   LINK_LIBS shaderc
+diff --git a/libshaderc_util/CMakeLists.txt b/libshaderc_util/CMakeLists.txt
+index 4926203..b152d7b 100644
+--- a/libshaderc_util/CMakeLists.txt
++++ b/libshaderc_util/CMakeLists.txt
+@@ -39,7 +39,10 @@ add_library(shaderc_util STATIC
+ 
+ shaderc_default_compile_options(shaderc_util)
+ target_include_directories(shaderc_util
+-  PUBLIC include PRIVATE ${glslang_SOURCE_DIR})
++  PUBLIC
++    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include/>
++    $<INSTALL_INTERFACE:include/>
++  PRIVATE ${glslang_SOURCE_DIR})
+ # We use parts of Glslang's HLSL compilation interface, which
+ # now requires this preprocessor definition.
+ add_definitions(-DENABLE_HLSL)
+@@ -50,12 +53,21 @@ target_link_libraries(shaderc_util PRIVATE
+   SPIRV-Tools-opt ${CMAKE_THREAD_LIBS_INIT})
+ 
+ if(SHADERC_ENABLE_INSTALL AND NOT BUILD_SHARED_LIBS)
+-  install(TARGETS shaderc_util
++  install(TARGETS shaderc_util EXPORT unofficial-shaderc_utilTargets
+     LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+     RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+     ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
++  install(EXPORT unofficial-shaderc_utilTargets
++      NAMESPACE unofficial::shaderc_util::
++      DESTINATION share/unofficial-shaderc_util
++  )
+ endif(SHADERC_ENABLE_INSTALL)
+ 
++file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/unofficial-shaderc_util-config.cmake.in"
++[[include("${CMAKE_CURRENT_LIST_DIR}/unofficial-shaderc_utilTargets.cmake")]])
++configure_file("${CMAKE_CURRENT_BINARY_DIR}/unofficial-shaderc_util-config.cmake.in" "${CMAKE_CURRENT_BINARY_DIR}/unofficial-shaderc_util-config.cmake" @ONLY)
++install(FILES ${CMAKE_CURRENT_BINARY_DIR}/unofficial-shaderc_util-config.cmake DESTINATION share/unofficial-shaderc_util)
++
+ shaderc_add_tests(
+   TEST_PREFIX shaderc_util
+   LINK_LIBS shaderc_util

--- a/ports/shaderc/portfile.cmake
+++ b/ports/shaderc/portfile.cmake
@@ -11,9 +11,10 @@ vcpkg_from_github(
         disable-update-version.patch
         fix-build-type.patch
         fix-install-shaderc_util.patch
+        fix-export-cmakefiles.patch
 )
 
-file(COPY ${CMAKE_CURRENT_LIST_DIR}/build-version.inc DESTINATION ${SOURCE_PATH}/glslc/src)
+file(COPY "${CMAKE_CURRENT_LIST_DIR}/build-version.inc" DESTINATION "${SOURCE_PATH}/glslc/src")
 
 set(OPTIONS)
 if(VCPKG_CRT_LINKAGE STREQUAL "dynamic")
@@ -25,9 +26,8 @@ vcpkg_find_acquire_program(PYTHON3)
 get_filename_component(PYTHON3_EXE_PATH ${PYTHON3} DIRECTORY)
 vcpkg_add_to_path(PREPEND "${PYTHON3_EXE_PATH}")
 
-vcpkg_configure_cmake(
-    SOURCE_PATH ${SOURCE_PATH}
-    PREFER_NINJA
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
     OPTIONS
         ${OPTIONS}
         -DSHADERC_SKIP_TESTS=true 
@@ -36,12 +36,14 @@ vcpkg_configure_cmake(
         -DSHADERC_ENABLE_EXAMPLES=OFF
 )
 
-vcpkg_install_cmake()
+vcpkg_cmake_install()
 
 vcpkg_fixup_pkgconfig()
+vcpkg_cmake_config_fixup(PACKAGE_NAME unofficial-shaderc CONFIG_PATH share/unofficial-shaderc)
+vcpkg_cmake_config_fixup(PACKAGE_NAME unofficial-shaderc_util CONFIG_PATH share/unofficial-shaderc_util)
 
 vcpkg_copy_tools(TOOL_NAMES glslc AUTO_CLEAN)
 
-file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
 
 file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)

--- a/ports/shaderc/vcpkg.json
+++ b/ports/shaderc/vcpkg.json
@@ -1,11 +1,20 @@
 {
   "name": "shaderc",
   "version": "2021.1",
-  "port-version": 1,
+  "port-version": 2,
   "description": "A collection of tools, libraries and tests for shader compilation.",
   "homepage": "https://github.com/google/shaderc",
+  "license": "Apache-2.0",
   "dependencies": [
     "glslang",
-    "spirv-tools"
+    "spirv-tools",
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    },
+    {
+      "name": "vcpkg-cmake-config",
+      "host": true
+    }
   ]
 }

--- a/versions/baseline.json
+++ b/versions/baseline.json
@@ -6366,7 +6366,7 @@
     },
     "shaderc": {
       "baseline": "2021.1",
-      "port-version": 1
+      "port-version": 2
     },
     "shaderwriter": {
       "baseline": "2.1.0",

--- a/versions/s-/shaderc.json
+++ b/versions/s-/shaderc.json
@@ -1,6 +1,11 @@
 {
   "versions": [
     {
+      "git-tree": "376570ffd232d9c4e0dad095ae3dea964dec52ed",
+      "version": "2021.1",
+      "port-version": 2
+    },
+    {
       "git-tree": "e8590ab232dd5ccd2331bb321127f00788b573a3",
       "version": "2021.1",
       "port-version": 1

